attelo/decoding/mst.py

```diff
@@ -1,88 +1,116 @@
+    """ Returns the root node for MST/MSDAG algorithm
 
+        Currently, the EDU in first position
+    """
+    return sorted(edus, key=lambda e:e.span)[0]
 
-def _graph(instances, root='1', use_prob=True):
-    """ instances are quadruplets of the form:
-
-            source, target, probability_of_attachment, relation
+def _graph(instances, use_prob=True):
+    """ Builds a directed graph for instances
 
-        root is the "root" of the graph, that is the node that has no incoming
-        nodes
+        instances are quadruplets of the form:
+            edu_source, edu_target, probability_of_attachment, relation
 
-        returns the Maximum Spanning Tree
+        returns a Digraph
     """
+    
+    root_id = _get_root(set(e for s,t,_,_ in instances for e in (s,t))).id
 
     targets = defaultdict(list)
     labels = dict()
     scores = dict()
 
     for source, target, prob, rel in instances:
-        src = source.id
-        tgt = target.id
-        if tgt == root:
+        src, tgt = source.id, target.id
+
+        # Ignore all edges directed to the root
+        if tgt == root_id:
             continue
-        scores[src, tgt] = prob
+
+        scores[src, tgt] = min(MAX_SCORE, max(MIN_SCORE, logit(prob))) if use_prob else prob
         labels[src, tgt] = rel
-        if use_prob:  # probability scores
-            scores[src, tgt] = log(prob if prob != 0.0 else sys.float_info.min)
         targets[src].append(tgt)
 
     return Digraph(targets,
                    lambda s, t: scores[s, t],
-                   lambda s, t: labels[s, t]).mst()
-
+                   lambda s, t: labels[s, t])
 
-def _list_edges(instances, root='1', use_prob=True):
-    """ instances are quadruplets of the form:
+def _msdag(graph):
+    """ Returns a subgraph of graph (a Digraph) corresponding to its
+        Maximum Spanning Directed Acyclic Graph
 
-            source, target, probability_of_attachment, relation
-
-        root is the "root" of the graph, that is the node that has no incoming
-        nodes
-
-        returns a list of edges for the MST graph
+        Alogrithm is semi-greedy-MSDAG as described in Schluter (2014)
     """
-    mst = _graph(instances, root, use_prob=use_prob)
-
-    return [(src, tgt, mst.get_label(src, tgt))
-            for src, tgt in mst.iteredges()]
-
+    tree = graph.mst()
+    # Sort edges in orginal graph by decreasing score
+    edges = sorted(graph.iteredges(), key=lambda p:-graph.get_score(*p))
+    for src, tgt in edges:
+        # Already in graph ?
+        if tgt in tree.successors[src]:
+            continue
+        # Add the edge, revert if cycle is created
+        tree.successors[src].append(tgt)
+        if tree.find_cycle():
+            tree.successors[src].remove(tgt)
+
+    # Update score and label functions
+    # I really wish that Digraph used a pair as single argument,
+    #   the following would be clearer
+    nscores, nlabels = (dict((k, old(*k)) for k in tree.iteredges())
```

Comment:
Alternatively, you might consider (at your option) the syntax `{k: old(*k) for k in tree.iteredges()}`

Note that pylint doesn't really like star args, although it may make sense to suppress the warning in this case.  pylint-clean code is much appreciated.